nni/retiarii/oneshot/pytorch/differentiable.py

```diff
@@ -0,0 +1,338 @@
+
+    def parameters(self):
+        for _, p in self.named_parameters():
+            yield p
+
+    def named_parameters(self, recurse):
+        for name, p in super(DartsInputChoice, self).named_parameters():
+            if name == 'alpha':
+                continue
+            yield name, p
+
+    def export(self):
+        return torch.argsort(-self.alpha).cpu().numpy().tolist()[:self.n_chosen]
+    
+    def r_forward(self, x):
+        return super().r_forward(x)
+
+class DartsModule(BaseOneShotLightningModule):
+    """
+    The DARTS model. In each iteration, there are 2 training phases. The phase 1 is architecture
+    step, in which the model parameters are frozen and the alphas are optimized. The phase 2 is model
+    step, in wchich the alphas are frozen and the model parameters are optimized. In this step, the
+    out put of each choice are sumed up with respect to their alpha value. The result of DARTS
+    is argmax in alpha.
+    The DARTS Model should be trained with ParallelTraiValDataloader in nn.retiarii.oneshot.pytorch.utils.
+    See base class for more attributes.
+
+    Reference
+    ----------
+    .. [darts] H. Liu, K. Simonyan, and Y. Yang, “DARTS: Differentiable Architecture Search,” presented at the
+        International Conference on Learning Representations, Sep. 2018. Available: https://openreview.net/forum?id=S1eYHoC5FX
+    """
+    
+    automatic_optimization = False
+    
+    def __init__(self, model, custom_replace_dict = None):
+        """
+        Parameters
+        ----------
+        base_model : pl.LightningModule
+            The module in evaluators in nni.retiarii.evaluator.lightning. User defined model
+            is wrapped by base_model, and base_model will be wrapped by this model.
+        custom_replace_dict : Dict[Type[nn.Module], Callable[[nn.Module], nn.Module]], default = None
+            The custom xxxChoice replace method. Keys should be xxxChoice type and values should 
+            return an nn.module. This custom replace dict will override the default replace
+            dict of each NAS method.
+        """
+        super().__init__(model, custom_replace_dict)
+    
+    def on_train_start(self) -> None:
+        return super().on_train_start()
+        
+    def training_step(self, batch, batch_idx):
+        # grad manually, only 1 architecture optimizer for darts
+        opts = self.optimizers()
+        arc_optim = opts[0]
+        w_optim = opts[1:]
+
+        # ParallelTrainValDataLoader will yield both train and val data in a batch
+        trn_batch, val_batch = batch
+
+        # phase 1: architecture step
+        # The resample_architecture hook is kept for some following darts-based NAS
+        # methods such as proxyless. See code of those methods for details. 
+        self.resample_architecture()
+        arc_optim.zero_grad()
+        arc_step_loss = self.model.training_step(val_batch, 2 * batch_idx)
```

Comment:
I think `training_step` also has multiple types of outputs.
https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#training-step